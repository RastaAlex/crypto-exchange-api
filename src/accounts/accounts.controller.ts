import { 
  Body,
  Controller,
  Get,
  Param,
  Post,
  BadRequestException,
  ParseIntPipe,
  HttpException,
} from '@nestjs/common';
import {
  ApiBadRequestResponse,
  ApiCreatedResponse,
  ApiNotFoundResponse,
  ApiOkResponse,
  ApiOperation,
  ApiTags,
} from '@nestjs/swagger';
import { AccountsService } from './accounts.service';
import { CryptoAccountDto } from './dtos/crypto-account.dto';

@ApiTags('Accounts')
@Controller('accounts')
export class AccountsController {
  constructor(private readonly accountsService: AccountsService) {}

  @Get()
  @ApiOkResponse({ type: [CryptoAccountDto], description: 'Get all accounts' })
  @ApiOperation({ summary: 'Get all accounts' })
  async getAllAccounts(): Promise<CryptoAccountDto[]> {
    return (await this.accountsService.getAllAccounts()).map(account => {
      const dto = new CryptoAccountDto();
      Object.assign(dto, account);
      return dto;
    });
  }

  @Get(':id')
  @ApiOkResponse({ type: CryptoAccountDto, description: 'Get account by id' })
  @ApiNotFoundResponse({ description: 'Account not found' })
  @ApiOperation({ summary: 'Get account by id' })
  async fetchAccountById(@Param('id', ParseIntPipe) id: number): Promise<CryptoAccountDto | null> {
    const account = await this.accountsService.getAccountById(id);
    if (!account) {
      throw new BadRequestException('Account not found');
    }

    const dto = new CryptoAccountDto();
    Object.assign(dto, account);
    return dto;
  }

  @Post()
  @ApiCreatedResponse({ type: CryptoAccountDto, description: 'Create a new account' })
  @ApiBadRequestResponse({ description: 'Invalid request data' })
  @ApiOperation({ summary: 'Create a new account' })
  async createAccount(@Body() dto: CryptoAccountDto): Promise<CryptoAccountDto> {
    try {
      const account = await this.accountsService.createAccount(dto);
      const responseDto = new CryptoAccountDto();
      Object.assign(responseDto, account);
      return responseDto;
    } catch (error) {
      if (error instanceof HttpException) {
        throw error;
      } else {
        throw new BadRequestException('Failed to create a new account');
      }
    }
  }
}
