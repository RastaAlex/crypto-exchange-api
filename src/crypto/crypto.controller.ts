import { BadRequestException, Controller, Get, Query } from '@nestjs/common';
import { CryptoService } from './crypto.service';
import {
  ApiBadRequestResponse,
  ApiOkResponse,
  ApiOperation,
  ApiQuery,
  ApiTags,
} from '@nestjs/swagger';

type ExchangeRateResponse = Record<string, number>;

@ApiTags('Crypto')
@Controller('crypto')
export class CryptoController {
  constructor(private readonly cryptoService: CryptoService) {}

  @Get('exchange-rate')
  @ApiOkResponse({ description: 'Get exchange rate for the specified crypto/fiat currency pair.' })
  @ApiBadRequestResponse({ description: 'Invalid crypto/fiat currency pair' })
  @ApiOperation({ summary: 'Get exchange rate for a crypto/fiat currency pair' })
  @ApiQuery({ name: 'from', description: 'Source currency' })
  @ApiQuery({ name: 'to', description: 'Target currency' })
  async getCurrencyExchangeRate(@Query('from') from: string, @Query('to') to: string): Promise<ExchangeRateResponse> {
    const pair = from.toUpperCase() + to.toUpperCase();
    const rate = this.cryptoService.getExchangeRate(pair);

    if (!rate) {
      throw new BadRequestException(`Exchange rate not found for pair: ${pair}`);
    }

    return {
      [pair]: rate,
    };
  }
}
