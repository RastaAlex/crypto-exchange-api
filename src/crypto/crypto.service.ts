import * as WebSocket from 'ws';
import { OnModuleInit } from '@nestjs/common';

export class CryptoService implements OnModuleInit {
  private ws: WebSocket;
  private exchangeRates: Record<string, number> = {};

  init() {
    this.ws = new WebSocket('wss://ws.kraken.com');

    this.ws.on('open', () => {
      this.ws.send(JSON.stringify({ event: 'subscribe', pair: ['BTC/USD', 'ETH/USD'], subscription: { name: 'ticker' } }));
    });

    this.ws.on('message', (data: WebSocket.Data) => {
      this.handleMessage(data);
    });
  }

  private handleMessage(data: WebSocket.Data) {
    const message = JSON.parse(data.toString());
    
    if (message[1] && message[1].c) {
      let pair = message[3].replace('/', '');

      if (pair === 'XBTUSD') {
        pair = 'BTCUSD';
      }

      const rate = parseFloat(message[1].c);
  
      this.exchangeRates[pair] = rate;
    }
}

  getExchangeRate(pair: string): number {
    return this.exchangeRates[pair];
  }

  async onModuleInit() {
    this.init();
  }
}