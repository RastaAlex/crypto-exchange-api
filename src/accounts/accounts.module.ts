import { Module } from '@nestjs/common';
import { AccountsController } from './accounts.controller';
import { AccountsService } from './accounts.service';
import { PrismaModule } from '@database/prisma.module';
import { CryptoModule } from '@crypto/crypto.module';

@Module({
  imports: [PrismaModule, CryptoModule],
  controllers: [AccountsController],
  providers: [AccountsService],
})
export class AccountsModule {}
