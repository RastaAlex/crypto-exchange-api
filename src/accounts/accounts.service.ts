import { Injectable, OnModuleInit } from '@nestjs/common';
import * as cron from 'node-cron';
import { PrismaService } from '@prisma/prisma.service';
import { CryptoService } from '@crypto/crypto.service';
import { Account } from '@prisma/client';
import { CreateAccountDto } from './dtos/create-account.dto';

@Injectable()
export class AccountsService implements OnModuleInit {
  constructor(
    private readonly prismaService: PrismaService,
    private readonly cryptoService: CryptoService,
  ) {}

  async onModuleInit() {
    cron.schedule('0 0 * * *', () => this.updateAccountBalances());
  }

  async createAccount(createAccountDto: CreateAccountDto): Promise<Account> {
    const { cryptoAsset, referenceCurrency, balanceInCryptoAsset } = createAccountDto;

    return await this.prismaService.account.create({
      data: {
        cryptoAsset,
        referenceCurrency,
        balanceInCryptoAsset,
        balanceInReferenceCurrency: 0,
      },
    });
  }

  async getAllAccounts(): Promise<Account[]> {
    return await this.prismaService.account.findMany();
  }

  async getAccountById(id: number): Promise<Account | null> {
    return await this.prismaService.account.findUnique({
      where: { id: Number(id) },
    });
  }

  private async updateAccountBalances() {
    const accounts = await this.getAllAccounts();

    for (const account of accounts) {
      const pair = account.cryptoAsset + account.referenceCurrency;
      const rate = this.cryptoService.getExchangeRate(pair);

      if (rate) {
        const updatedBalance = account.balanceInCryptoAsset * rate;

        await this.prismaService.account.update({
          where: { id: account.id },
          data: { balanceInReferenceCurrency: updatedBalance },
        });
      }
    }
  }
}
