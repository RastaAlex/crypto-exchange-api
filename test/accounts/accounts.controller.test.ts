import { Test, TestingModule } from '@nestjs/testing';
import { AccountsController } from '@accounts/accounts.controller';
import { AccountsService } from '@accounts/accounts.service';
import { PrismaService } from '@database/prisma.service';
import { CryptoService } from '@crypto/crypto.service';
import { CreateAccountDto } from '@accounts/dtos/create-account.dto';

describe('AccountsController', () => {
  let accountsController: AccountsController;
  let accountsService: AccountsService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      controllers: [AccountsController],
      providers: [AccountsService, PrismaService, CryptoService],
    }).compile();
  
    accountsController = module.get<AccountsController>(AccountsController);
    accountsService = module.get<AccountsService>(AccountsService);
  });

  it('should be defined', () => {
    expect(accountsController).toBeDefined();
  });

  describe('getAllAccounts', () => {
    it('should return an array of accounts', async () => {
      const mockAccounts = [
        {
          id: 1,
          cryptoAsset: 'BTC',
          referenceCurrency: 'USD',
          balanceInCryptoAsset: 1,
          balanceInReferenceCurrency: 10,
        },
      ];

      jest.spyOn(accountsService, 'getAllAccounts').mockImplementation(async () => mockAccounts);

      const accounts = await accountsController.getAllAccounts();

      expect(accountsService.getAllAccounts).toHaveBeenCalled();
      expect(accounts).toMatchSnapshot();
    });
  });

  describe('getAccountById', () => {
    it('should return an account with the specified id', async () => {
      const mockAccount = {
        id: 1,
        cryptoAsset: 'BTC',
        referenceCurrency: 'USD',
        balanceInCryptoAsset: 1,
        balanceInReferenceCurrency: 10,
      };

      jest.spyOn(accountsService, 'getAccountById').mockImplementation(async () => mockAccount);

      const account = await accountsController.getAccountById(1);

      expect(accountsService.getAccountById).toHaveBeenCalledWith(1);
      expect(account).toMatchSnapshot();
    });

    it('should return null if the account is not found', async () => {
      jest.spyOn(accountsService, 'getAccountById').mockImplementation(async () => null);

      const account = await accountsController.getAccountById(42);

      expect(accountsService.getAccountById).toHaveBeenCalledWith(42);
      expect(account).toMatchSnapshot();
    });
  });

  describe('createAccount', () => {
    it('should create and return a new account', async () => {
      const createAccountDto: CreateAccountDto = {
        cryptoAsset: 'BTC',
        referenceCurrency: 'USD',
        balanceInCryptoAsset: 1,
      };

      const mockAccount = {
        id: 1,
        cryptoAsset: 'BTC',
        referenceCurrency: 'USD',
        balanceInCryptoAsset: 1,
        balanceInReferenceCurrency: 10,
      };

      jest.spyOn(accountsService, 'createAccount').mockImplementation(async () => mockAccount);

      const account = await accountsController.createAccount(createAccountDto);

      expect(accountsService.createAccount).toHaveBeenCalledWith(createAccountDto);
      expect(account).toMatchSnapshot();
    });
  });
});