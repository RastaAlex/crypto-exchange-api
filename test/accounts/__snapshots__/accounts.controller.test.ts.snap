// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountsController createAccount should create and return a new account 1`] = `
{
  "balanceInCryptoAsset": 1,
  "balanceInReferenceCurrency": 10,
  "cryptoAsset": "BTC",
  "id": 1,
  "referenceCurrency": "USD",
}
`;

exports[`AccountsController getAccountById should return an account with the specified id 1`] = `
{
  "balanceInCryptoAsset": 1,
  "balanceInReferenceCurrency": 10,
  "cryptoAsset": "BTC",
  "id": 1,
  "referenceCurrency": "USD",
}
`;

exports[`AccountsController getAccountById should return null if the account is not found 1`] = `null`;

exports[`AccountsController getAllAccounts should return an array of accounts 1`] = `
[
  {
    "balanceInCryptoAsset": 1,
    "balanceInReferenceCurrency": 10,
    "cryptoAsset": "BTC",
    "id": 1,
    "referenceCurrency": "USD",
  },
]
`;
